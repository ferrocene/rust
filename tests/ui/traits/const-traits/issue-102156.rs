//@ revisions: edition2015 edition2018
//@ [edition2015] edition: 2015
//@ [edition2018] edition: 2018..

#![feature(allocator_api)]
#![feature(const_trait_impl)]

use core::convert::{From, TryFrom};
//[edition2015]~^ ERROR failed to resolve: you might be missing crate `core`
//[edition2015]~| ERROR failed to resolve: you might be missing crate `core`

use std::pin::Pin;
use std::alloc::Allocator;
impl<T: ?Sized, A: Allocator> const From<Box<T, A>> for Pin<Box<T, A>>
//[edition2018]~^ ERROR const `impl` for trait `From` which is not marked with `#[const_trait]`
//[edition2018]~| ERROR type parameter `T` must be used as the type parameter for some local type
where
    A: 'static,
{}

pub fn main() {}
