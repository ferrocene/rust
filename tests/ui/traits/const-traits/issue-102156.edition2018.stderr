error: const `impl` for trait `From` which is not marked with `#[const_trait]`
  --> $DIR/issue-102156.rs:14:37
   |
LL | impl<T: ?Sized, A: Allocator> const From<Box<T, A>> for Pin<Box<T, A>>
   |                                     ^^^^^^^^^^^^^^^ this trait is not `const`
   |
   = note: marking a trait with `#[const_trait]` ensures all default method bodies are `const`
   = note: adding a non-const method body in the future would be a breaking change

error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
  --> $DIR/issue-102156.rs:14:6
   |
LL | impl<T: ?Sized, A: Allocator> const From<Box<T, A>> for Pin<Box<T, A>>
   |      ^ type parameter `T` must be used as the type parameter for some local type
   |
   = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local
   = note: only traits defined in the current crate can be implemented for a type parameter

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0210`.
